dev-libs/tree-sitter/tree-sitter-0.20.1-r1.ebuild

```diff
@@ -0,0 +1,172 @@
+# Copyright 1999-2021 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=8
+
+CARGO_OPTIONAL=1
+
+CRATES="
+	aho-corasick-0.7.15
+	ansi_term-0.11.0
+	ansi_term-0.12.1
+	anyhow-1.0.40
+	arrayref-0.3.6
+	arrayvec-0.5.2
+	ascii-1.0.0
+	atty-0.2.14
+	autocfg-1.0.1
+	base64-0.13.0
+	bitflags-1.2.1
+	blake2b_simd-0.5.11
+	bumpalo-3.6.1
+	cc-1.0.67
+	cfg-if-1.0.0
+	chrono-0.4.19
+	chunked_transfer-1.4.0
+	clap-2.33.3
+	constant_time_eq-0.1.5
+	crossbeam-utils-0.8.3
+	ctor-0.1.20
+	diff-0.1.12
+	difference-2.0.0
+	dirs-3.0.1
+	dirs-sys-0.3.5
+	either-1.6.1
+	form_urlencoded-1.0.1
+	getrandom-0.1.16
+	getrandom-0.2.2
+	glob-0.3.0
+	hashbrown-0.9.1
+	hermit-abi-0.1.18
+	html-escape-0.2.6
+	idna-0.2.2
+	indexmap-1.6.1
+	itoa-0.4.7
+	js-sys-0.3.48
+	lazy_static-1.4.0
+	libc-0.2.86
+	libloading-0.7.0
+	log-0.4.14
+	matches-0.1.8
+	memchr-2.3.4
+	num-integer-0.1.44
+	num-traits-0.2.14
+	once_cell-1.7.0
+	output_vt100-0.1.2
+	percent-encoding-2.1.0
+	ppv-lite86-0.2.10
+	pretty_assertions-0.7.2
+	proc-macro2-1.0.24
+	quote-1.0.9
+	rand-0.8.3
+	rand_chacha-0.3.0
+	rand_core-0.6.2
+	rand_hc-0.3.0
+	redox_syscall-0.1.57
+	redox_syscall-0.2.5
+	redox_users-0.3.5
+	regex-1.4.3
+	regex-syntax-0.6.22
+	remove_dir_all-0.5.3
+	rust-argon2-0.8.3
+	rustc-hash-1.1.0
+	ryu-1.0.5
+	same-file-1.0.6
+	serde-1.0.130
+	serde_derive-1.0.130
+	serde_json-1.0.63
+	smallbitvec-2.5.1
+	spin-0.7.1
+	strsim-0.8.0
+	syn-1.0.67
+	tempfile-3.2.0
+	textwrap-0.11.0
+	thiserror-1.0.25
+	thiserror-impl-1.0.25
+	thread_local-1.1.3
+	time-0.1.43
+	tiny_http-0.8.0
+	tinyvec-1.1.1
+	tinyvec_macros-0.1.0
+	toml-0.5.8
+	unicode-bidi-0.3.4
+	unicode-normalization-0.1.17
+	unicode-width-0.1.8
+	unicode-xid-0.2.1
+	url-2.2.1
+	utf8-width-0.1.4
+	vec_map-0.8.2
+	walkdir-2.3.1
+	wasi-0.9.0+wasi-snapshot-preview1
+	wasi-0.10.2+wasi-snapshot-preview1
+	wasm-bindgen-0.2.71
+	wasm-bindgen-backend-0.2.71
+	wasm-bindgen-macro-0.2.71
+	wasm-bindgen-macro-support-0.2.71
+	wasm-bindgen-shared-0.2.71
+	web-sys-0.3.48
+	webbrowser-0.5.5
+	which-4.1.0
+	widestring-0.4.3
+	winapi-0.3.9
+	winapi-i686-pc-windows-gnu-0.4.0
+	winapi-util-0.1.5
+	winapi-x86_64-pc-windows-gnu-0.4.0
+"
+
+inherit toolchain-funcs cargo
+
+DESCRIPTION="Tree-sitter is a parser generator tool and an incremental parsing library."
+HOMEPAGE="https://github.com/tree-sitter/tree-sitter"
+
+if [[ ${PV} == *9999* ]]; then
+	inherit git-r3
+	EGIT_REPO_URI="https://github.com/${PN}/${PN}"
+else
+	SRC_URI="
+		https://github.com/${PN}/${PN}/archive/${PV}.tar.gz -> ${P}.tar.gz
+		ts-cli? ( $(cargo_crate_uris) )
+	"
+	KEYWORDS="~amd64"
+fi
+
+LICENSE="Apache-2.0 BSD-2 CC0-1.0 ISC MIT"
```

Comment:
If the non-MIT licenses are from the cli tool stuff, we should probably have a conditiona license

LICENSE="MIT ts-cli? ( Apache-2.0 BSD-2 CC0-1.0 ISC )" or whatever

Reply:
Yeah, we discussed that on IRC already.
I've already changed that locally, just we didn't find final conclusion about how that should look like to properly reflect the "OR"'s in crates licenses.

The thing is in that actually `cargo licanse` says that:

```
Apache-2.0 (1): chunked_transfer
Apache-2.0 OR Apache-2.0 WITH LLVM-exception OR MIT (2): wasi, wasi
Apache-2.0 OR BSL-1.0 (1): ryu
Apache-2.0 OR MIT (75): anyhow, arrayvec, ascii, autocfg, base64, bitflags, bumpalo, cc, cfg-if, chrono, crossbeam-utils, ctor, diff, dirs, dirs-sys, either, form_urlencoded, getrandom, getrandom, glob, hashbrown, hermit-abi, idna, indexmap, itoa, js-sys, lazy_static, libc, log, num-integer, num-traits, once_cell, percent-encoding, ppv-lite86, pretty_assertions, proc-macro2, quote, rand, rand_chacha, rand_core, rand_hc, regex, regex-syntax, remove_dir_all, rust-argon2, rustc-hash, serde, serde_derive, serde_json, smallbitvec, syn, tempfile, thiserror, thiserror-impl, thread_local, time, tiny_http, toml, unicode-bidi, unicode-normalization, unicode-width, unicode-xid, url, vec_map, wasm-bindgen, wasm-bindgen-backend, wasm-bindgen-macro, wasm-bindgen-macro-support, wasm-bindgen-shared, web-sys, webbrowser, widestring, winapi, winapi-i686-pc-windows-gnu, winapi-x86_64-pc-windows-gnu
Apache-2.0 OR MIT OR Zlib (2): tinyvec, tinyvec_macros
BSD-2-Clause (1): arrayref
CC0-1.0 (1): constant_time_eq
ISC (1): libloading
MIT (23): ansi_term, ansi_term, atty, blake2b_simd, clap, difference, html-escape, matches, output_vt100, redox_syscall, redox_syscall, redox_users, spin, strsim, textwrap, tree-sitter, tree-sitter-cli, tree-sitter-config, tree-sitter-highlight, tree-sitter-loader, tree-sitter-tags, utf8-width, which
MIT OR Unlicense (5): aho-corasick, memchr, same-file, walkdir, winapi-util
```

And all that I can imagine to reflect that is:
```
LICENSE="
        MIT
        ts-cli? (
                Apache-2.0
                || ( Apache-2.0 OR Apache-2.0-with-LLVM-exceptions MIT )
                || ( Apache-2.0 Boost )
                || ( Apache-2.0 MIT )
                || ( Apache-2.0 MIT ZLIB )
                BSD-2
                CC0-1.0
                ISC
                MIT
                || ( MIT Unlicense )
        )
"
```

But I'm unsure about such amount of duplication.

And I also unsure if
```
LICENSE="
  MIT
  ts-cli? ( Apache-2.0 BSD-2 CC0-1.0 ISC MIT )
"
```

would suffice...

And what do you think about it?

Reply:
Looks correct to me